agent/eventhandler/task_handler.go

```diff
@@ -15,244 +15,323 @@ package eventhandler
-	sending bool
-	//eventsListLock locks both the list and sending bool
-	eventListLock sync.Mutex
-}
+	// drainEventsFrequency is the frequency at the which unsent events batched
+	// by the task handler are sent to the backend
+	drainEventsFrequency = time.Minute
+
+	submitStateBackoffMin            = time.Second
+	submitStateBackoffMax            = 30 * time.Second
+	submitStateBackoffJitterMultiple = 0.20
+	submitStateBackoffMultiple       = 1.3
+)
 
 // TaskHandler encapsulates the the map of a task arn to task and container events
 // associated with said task
 type TaskHandler struct {
 	// submitSemaphore for the number of tasks that may be handled at once
 	submitSemaphore utils.Semaphore
 	// taskToEvents is arn:*eventList map so events may be serialized per task
-	//TODO: fix leak, currently items are never removed from this map
-	tasksToEvents map[string]*eventList
+	tasksToEvents map[string]*taskSendableEvents
 	// tasksToContainerStates is used to collect container events
 	// between task transitions
 	tasksToContainerStates map[string][]api.ContainerStateChange
 
 	//  taskHandlerLock is used to safely access the following maps:
 	// * taskToEvents
 	// * tasksToContainerStates
-	taskHandlerLock sync.RWMutex
+	lock sync.RWMutex
 
 	// stateSaver is a statemanager which may be used to save any
 	// changes to a task or container's SentStatus
 	stateSaver statemanager.Saver
+
+	drainEventsFrequency time.Duration
+	state                dockerstate.TaskEngineState
+	client               api.ECSClient
+	ctx                  context.Context
+}
+
+// taskSendableEvents is used to group all events for a task
+type taskSendableEvents struct {
+	// events is a list of *sendableEvents
+	events *list.List
+	// sending will check whether the list is already being handlerd
+	sending bool
+	//eventsListLock locks both the list and sending bool
+	lock sync.Mutex
+	// createdAt is a timestamp for when the event list was created
+	createdAt time.Time
+	// taskARN is the task arn that the event list is associated with
+	taskARN string
 }
 
 // NewTaskHandler returns a pointer to TaskHandler
-func NewTaskHandler(stateManager statemanager.Saver) *TaskHandler {
-	return &TaskHandler{
-		tasksToEvents:          make(map[string]*eventList),
+func NewTaskHandler(ctx context.Context,
+	stateManager statemanager.Saver,
+	state dockerstate.TaskEngineState,
+	client api.ECSClient) *TaskHandler {
+	// Create a handler and start the periodic event drain loop
+	taskHandler := &TaskHandler{
+		ctx:                    ctx,
+		tasksToEvents:          make(map[string]*taskSendableEvents),
 		submitSemaphore:        utils.NewSemaphore(concurrentEventCalls),
 		tasksToContainerStates: make(map[string][]api.ContainerStateChange),
 		stateSaver:             stateManager,
+		state:                  state,
+		client:                 client,
+		drainEventsFrequency:   drainEventsFrequency,
 	}
+	go taskHandler.startDrainEventsTicker()
```

Comment:
I'd like a comment somewhere explaining the flow for batched events.  This is roughly what I got through reading this:

```
startDrainEventsTicker->
	getBatchedContainerEvents->
		read from tasksToContainerStates and returns TaskEvent
	AddStateChangeEvent->
		flushBatchUnsafe
```

The part that took me a while is that `AddStateChangeEvent` also calls `batchContainerUnsafe` for `ContainerEvent`s and the key difference is that `getBatchedContainerEvents` returns a `TaskEvent` even though it has "ContainerEvents" in the name.

Reply:
I'll rename `getBatchedContainerEvents` -> `taskStateChangesToSend`. Will add some comments in `startDrainEventsTicker` as well.